### PR TITLE
win: restore file position after positional read/write operation

### PR DESCRIPTION
**Pull request by bzoz**
File read or write from specified position will move file pointer on Windows but not on POSIX. This makes Windows behave as other supported platforms: for writes/reads with provided offset, file position before operation is saved, and then restored once the operation completes.

Ref: https://github.com/nodejs/node/issues/9671

### CODE REVIEW

**Comment by sam-github** [comment]
I've noticed this before, it surfaces as an undocumented aspect of the node fs.read/write calls. The windows implementation can't be atomic, as `pread()`/`pwrite()` are, but I don't think that matters, and it would be good to bring Unix/Windows into alignment here.

File: src/win/fs.c [hunk 1]
```diff
@@ -593,6 +602,10 @@ void fs__read(uv_fs_t* req) {
     ++index;
   } while (result && index < req->fs.info.nbufs);
 
+  if (restore_position) {
```
Comment (saghul) [inline src/win/fs.c]: style: drop braces here

File: src/win/fs.c [hunk 2]
```diff
@@ -651,6 +673,10 @@ void fs__write(uv_fs_t* req) {
...
+    SetFilePointerEx(handle, original_position, NULL, FILE_BEGIN);
```
Comment (saghul) [inline src/win/fs.c]: ditto

File: test/test-fs.c [hunk 3]
```diff
@@ -2765,3 +2765,44 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
 }
+
+TEST_IMPL(fs_file_pos_after_op_with_offset) {
+  int r;
+
+  /* Setup. */
+  unlink("test_file");
+  loop = uv_default_loop();
+
+  r = uv_fs_open(loop, &open_req1, "test_file", O_RDWR | O_CREAT,
```
Comment (saghul) [inline test/test-fs.c]: style: please line arguments vertically

File: test/test-fs.c [hunk 4]
```diff
@@ -2765,3 +2765,44 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+      S_IWUSR | S_IRUSR, NULL);
+  ASSERT(r >= 0);
+  ASSERT(open_req1.result >= 0);
```
Comment (saghul) [inline test/test-fs.c]: this assert is kinda redundant, since the result of a sync open is the same as the return value.

File: test/test-fs.c [hunk 5]
```diff
@@ -2765,3 +2765,44 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+  uv_fs_req_cleanup(&open_req1);
+
+  iov = uv_buf_init(test_buf, sizeof(test_buf));
+  r = uv_fs_write(NULL, &write_req, open_req1.result, &iov, 1, 0, NULL);
+  ASSERT(r >= 0);
+  ASSERT(write_req.result >= 0);
```
Comment (saghul) [inline test/test-fs.c]: ditto

File: test/test-fs.c [hunk 6]
```diff
@@ -2765,3 +2765,44 @@ TEST_IMPL(fs_read_write_null_arguments) {
...
+  ASSERT(lseek(open_req1.result, 0, SEEK_CUR) == 0);
+  uv_fs_req_cleanup(&write_req);
+
+  iov = uv_buf_init(buf, sizeof(buf));
+  r = uv_fs_read(NULL, &read_req, open_req1.result, &iov, 1, 0, NULL);
+  ASSERT(r >= 0);
+  ASSERT(read_req.result >= 0);
```
Comment (saghul) [inline test/test-fs.c]: ditto

**Comment by saghul** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/279/

**Comment by bzoz** [comment]
Updated, PTAL

**Comment by bzoz** [comment]
Rebased, PTAL

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/285/

**Comment by bzoz** [comment]
Ping?

File: test/test-fs.c [hunk 7]
```diff
@@ -2804,3 +2804,43 @@ TEST_IMPL(get_osfhandle_valid_handle) {
...
+  unlink("test_file");
+  loop = uv_default_loop();
+
+  r = uv_fs_open(loop,
+                 &open_req1,
+                 "test_file",
+                 O_RDWR | O_CREAT,
+                 S_IWUSR | S_IRUSR,
+                 NULL);
+  ASSERT(r >= 0);
```
Comment (santigimeno) [inline test/test-fs.c]: Shouldn't `r` be greater than `0`?

File: test/test-fs.c [hunk 8]
```diff
@@ -2804,3 +2804,43 @@ TEST_IMPL(get_osfhandle_valid_handle) {
...
+  uv_fs_req_cleanup(&open_req1);
+
+  iov = uv_buf_init(test_buf, sizeof(test_buf));
+  r = uv_fs_write(NULL, &write_req, open_req1.result, &iov, 1, 0, NULL);
+  ASSERT(r >= 0);
```
Comment (santigimeno) [inline test/test-fs.c]: shouldn't we check the number `r == sizeof(test_buf)`?

File: test/test-fs.c [hunk 9]
```diff
@@ -2804,3 +2804,43 @@ TEST_IMPL(get_osfhandle_valid_handle) {
...
+  ASSERT(lseek(open_req1.result, 0, SEEK_CUR) == 0);
+  uv_fs_req_cleanup(&write_req);
+
+  iov = uv_buf_init(buf, sizeof(buf));
+  r = uv_fs_read(NULL, &read_req, open_req1.result, &iov, 1, 0, NULL);
+  ASSERT(r >= 0);
```
Comment (santigimeno) [inline test/test-fs.c]: also here?

File: src/win/fs.c [hunk 10]
```diff
@@ -627,6 +644,10 @@ void fs__write(uv_fs_t* req) {
   if (offset != -1) {
     memset(&overlapped, 0, sizeof overlapped);
     overlapped_ptr = &overlapped;
+    if (SetFilePointerEx(handle, zero_offset, &original_position,
+                         FILE_CURRENT)) {
+      restore_position = 1;
```
Comment (santigimeno) [inline src/win/fs.c]: On which cases can this fail? Should we note in the documentation that on Windows the offset could actually be modified after calling `uv_fs_write() / uv_fs_read()`?

File: test/test-fs.c (hunk 7 above)
Comment (bzoz) [inline test/test-fs.c]: My test is modified `fs_file_open_append` test, didn't give a second though. I'll update the tests.

File: src/win/fs.c (hunk 10 above)
Comment (bzoz) [inline src/win/fs.c]: As per [MSDN SetFilePointerEx doc](https://msdn.microsoft.com/pl-pl/library/windows/desktop/aa365542.aspx) this call can fail if the handle is not a file or if it was opened with `FILE_FLAG_NO_BUFFERING`. We test handle type and libuv does not uses that flag, so I guess we are safe.

**Comment by bzoz** [comment]
Updated, PTAL

**Comment by bzoz** [comment]
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/291/

**Comment by bzoz** [comment]
@saghul, can I get +1 from you?

**Comment by bzoz** [comment]
Landed in https://github.com/libuv/libuv/commit/0bd8f5bf36b4164b071af6b0bb68c814cf7dedf8

**Pull request closed without merging**